openedx/features/subscriptions/models.py

```diff
@@ -0,0 +1,104 @@
+    """
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
+    subscription_id = models.PositiveIntegerField(db_index=True)
+    expiration_date = models.DateField(default=None, null=True, blank=True)
+
+    LIMITED_ACCESS = 'limited-access'
+    FULL_ACCESS_COURSES = 'full-access-courses'
+    FULL_ACCESS_TIME_PERIOD = 'full-access-time-period'
+    LIFETIME_ACCESS = 'lifetime-access'
+
+    SUBSCRIPTION_TYPE_CHOICES = (
+        (LIMITED_ACCESS, 'Limited Access', ),
+        (FULL_ACCESS_COURSES, 'Full Access (Courses)', ),
+        (FULL_ACCESS_TIME_PERIOD, 'Full Access (Time period)', ),
+        (LIFETIME_ACCESS, 'Lifetime Access', ),
+    )
+
+    subscription_type = models.CharField(choices=SUBSCRIPTION_TYPE_CHOICES, max_length=30)
+    max_allowed_courses = models.PositiveIntegerField(
+        help_text=_(
+            'Maximum number of courses allowed to enroll in paid tracks using the subscription.'
+        ),
+        null=True,
+        blank=True
+    )
+    course_enrollments = models.ManyToManyField(CourseEnrollment, blank=True)
+
+    def __str__(self):
+        return "Subscription of type {subscription_type} with ID {subscription_id} for User {user_email}".format(
+            subscription_type=self.subscription_type,
+            subscription_id=self.subscription_id,
+            user_email=self.user.username,
+        )
+
+    @property
+    def is_active(self):
+        """
+        Get active status of a subscription.
+
+        Check if a subscription is active even if a user can no longer purchase a new course through it.
+        """
+        course_enrollments_count = self.course_enrollments.count()
+        if self.expiration_date and self.max_allowed_courses:
+            return self.expiration_date >= date.today() and course_enrollments_count <= self.max_allowed_courses
+        elif self.expiration_date and not self.max_allowed_courses:
+            return self.expiration_date >= date.today()
+        elif self.max_allowed_courses and not self.expiration_date :
+            return course_enrollments_count <= self.max_allowed_courses
+
+        return True
+
+    @property
+    def is_valid(self):
+        """
+        Get validity status of a subscription.
+
+        Check if a user can enroll in a new course through this subscription.
+        """
+        course_enrollments_count = self.course_enrollments.count()
+        if self.subscription_type == self.LIFETIME_ACCESS:
+            return True
+        elif self.subscription_type == self.FULL_ACCESS_COURSES:
+            return self.expiration_date >= date.today()
+        elif self.subscription_type == self.FULL_ACCESS_TIME_PERIOD:
+            return course_enrollments_count < self.max_allowed_courses
+        elif self.subscription_type == self.LIMITED_ACCESS:
+            return self.expiration_date >= date.today() and course_enrollments_count < self.max_allowed_courses
+
+    @classmethod
+    def get_valid_subscription(self, user_id=None):
```

Comment:
Since this method is now returning a queryset (a list of items) so now we need to update its name `get_valid_subscriptions` and docstring `Get valid subscription of a user.`:
```python
def get_valid_subscriptions(self, user_id=None):        """
        Get valid subscriptions of a user.

        Returns:
            queryset: Returns latest valid user subscription.
        """
```

Reply:
done